File: 279. Perfect Squares/Solution.py
class Solution:
    def numSquares(self, n: int) -> int:
        if n <= 0:
            return 0
        
        perfect_squares = [i * i for i in range(1, int(n ** 0.5) + 1)]
        
        queue = deque([(n, 0)])
        visited = set()
        
        while queue:
            num, level = queue.popleft()
            for square in perfect_squares:
                diff = num - square
                if diff == 0:
                    return level + 1
                elif diff > 0 and diff not in visited:
                    queue.append((diff, level + 1))
                    visited.add(diff)
        
        return -1  # If no solution found

File: 279. Perfect Squares/Solution.cpp
class Solution {
public:
    int numSquares(int n) {
        if (n <= 0) return 0;
        
        // Generate perfect squares
        vector<int> perfectSquares;
        for (int i = 1; i * i <= n; ++i) {
            perfectSquares.push_back(i * i);
        }
        
        // BFS
        queue<pair<int, int>> q;
        q.push({n, 0});
        vector<bool> visited(n + 1, false);
        visited[n] = true;
        
        while (!q.empty()) {
            int num = q.front().first;
            int level = q.front().second;
            q.pop();
            
            for (int square : perfectSquares) {
                int diff = num - square;
                if (diff == 0) {
                    return level + 1;
                } else if (diff > 0 && !visited[diff]) {
                    q.push({diff, level + 1});
                    visited[diff] = true;
                }
            }
        }
        
        return -1; // If no solution found
    }
};